### PR TITLE
Add workspace filter/column into saved objects page

### PR DESCRIPTION
**Pull request by Hailong-am**
### Description
1. Add workspace filter into saved objects page
2. Add workspace column into saved objects page

**workspace column**
image

**workspace filter includes default workspace**
image

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx [hunk 1]
```diff
@@ -231,9 +236,15 @@ export class Table extends PureComponent<TableProps, TableState> {
           if (!canGoInApp) {
             return <EuiText size="s">{title || getDefaultTitle(object)}</EuiText>;
           }
-          return (
-            <EuiLink href={basePath.prepend(path)}>{title || getDefaultTitle(object)}</EuiLink>
-          );
+          let inAppUrl = basePath.prepend(path);
+          if (object.workspaces?.length) {
+            // first workspace login user have permission
+            const [workspaceId] = object.workspaces.filter((wsId) => visibleWsIds.includes(wsId));
```
Comment (wanglam) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:242]: Do we need to consider use currentWorkspaceId here?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx [hunk 2]
```diff
@@ -571,5 +581,145 @@ describe('SavedObjectsTable', () => {
...
+      await new Promise((resolve) => process.nextTick(resolve));
+      // Ensure the state changes are reflected
+      component.update();
+
+      const props = component.find('Table').props() as TableProps;
+      const filters = props.filters;
+      const wsFilter = filters.filter((f) => f.field === 'workspaces');
+      expect(wsFilter.length).toBe(1);
+      expect(wsFilter[0].options.length).toBe(1);
+      expect(wsFilter[0].options[0].value).toBe('foo');
```
Comment (wanglam) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx:672]: What is the difference of above `show workspace filter when workspace on and not in a workspace` test case?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 3]
```diff
@@ -245,31 +304,52 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
     this.setState({ isSearching: true }, this.debouncedFetchObjects);
   };
 
+  fetchWorkspace = () => {
```
Comment (wanglam) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:307]: What about renaming to subscribe workspace changes?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 4]
```diff
@@ -166,37 +177,85 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
...
+      workspaces: this.workspaceIdQuery,
+    });
 
     if (availableNamespaces.length) {
       const filteredNamespaces = filterQuery(availableNamespaces, visibleNamespaces);
       filteredCountOptions.namespacesToInclude = filteredNamespaces;
     }
+    if (visibleWorkspaces?.length) {
+      filteredCountOptions.workspaces = visibleWorkspaces.map(
+        (wsName) => this.wsNameIdLookup?.get(wsName) || ''
```
Comment (wanglam) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:256]: Do we need to filter out empty workspace id?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx (hunk 2 above)
Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx:672]: when in a workspace, the filter have only one option which is current workspace. 
when not in a workspace, the filter will have all visible workspace as option

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx (hunk 1 above)
Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:242]: you are right, add current workspace logic here

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx (hunk 3 above)
Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:307]: make sense, updated

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx (hunk 4 above)
Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:256]: sounds good, add filter logic

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx [hunk 5]
```diff
@@ -62,6 +62,10 @@ import {
 } from './saved_objects_table';
 import { Flyout, Relationships } from './components';
 import { SavedObjectWithMetadata } from '../../types';
+import { WorkspaceObject } from 'opensearch-dashboards/public';
+import { DEFAULT_WORKSPACE_ID } from '../../../../../core/public';
+import { TableProps } from './components/table';
+import { waitFor } from '@testing-library/dom';
```
Comment (SuZhou-Joe) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.test.tsx:68]: These lines should be linted when commit(Absolute import should move before relative import), could you please take a look to see if the lint take effect when you do git commit?

File: src/core/utils/constants.ts [hunk 6]
```diff
@@ -8,3 +8,5 @@ export const WORKSPACE_TYPE = 'workspace';
 export const WORKSPACE_PATH_PREFIX = '/w';
 
 export const PUBLIC_WORKSPACE_ID = 'public';
+
+export const DEFAULT_WORKSPACE_ID = 'default';
```
Comment (SuZhou-Joe) [inline src/core/utils/constants.ts:12]: In 2.13 I thought `default workspace` is a virtual workspace? Could we add some comment on it as at first I thought it is a new physical workspace.

File: src/plugins/workspace/public/plugin.test.ts [hunk 7]
```diff
@@ -9,6 +9,7 @@ import { applicationServiceMock, chromeServiceMock, coreMock } from '../../../co
 import { WorkspacePlugin } from './plugin';
 import { WORKSPACE_FATAL_ERROR_APP_ID, WORKSPACE_OVERVIEW_APP_ID } from '../common/constants';
 import { Observable, Subscriber } from 'rxjs';
+import { savedObjectsManagementPluginMock } from 'src/plugins/saved_objects_management/public/mocks';
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/plugin.test.ts:12]: Perhaps we need to change to a relative path.

**Review by SuZhou-Joe — APPROVED** [review]

File: src/core/utils/constants.ts (hunk 6 above)
Comment (Hailong-am) [reply src/core/utils/constants.ts:16]: sounds good, comments added.

File: src/plugins/workspace/public/plugin.test.ts (hunk 7 above)
Comment (Hailong-am) [reply src/plugins/workspace/public/plugin.test.ts]: thanks, changed to relative import.

**Pull request merged**